File: backend/internal/domain/money.go
package domain

import "strconv"

// FormatIDR formats whole-rupiah int64 amounts as "Rp1.234.567,-" - the
// canonical app-wide Rupiah nominal (matches the frontend $lib/money.formatIDR).
// Negative amounts render as "-Rp1.234.567,-".
func FormatIDR(amount int64) string {
	neg := amount < 0
	if neg {
		amount = -amount
	}
	digits := strconv.FormatInt(amount, 10)

	// Insert '.' thousands separators from the right, then a trailing ",-".
	n := len(digits)
	groups := (n - 1) / 3
	out := make([]byte, 0, n+groups+5)
	if neg {
		out = append(out, '-')
	}
	out = append(out, 'R', 'p')
	lead := n - groups*3
	out = append(out, digits[:lead]...)
	for i := lead; i < n; i += 3 {
		out = append(out, '.')
		out = append(out, digits[i:i+3]...)
	}
	out = append(out, ',', '-')
	return string(out)
}

File: frontend/src/lib/money.ts
/**
 * Canonical Rupiah nominal format for the whole app: `Rp1.000.000,-`.
 *
 * No space after "Rp", "." as thousands separator (id-ID locale), whole rupiah
 * only (money is int64 IDR - CONTRACTS.md §3), and a trailing ",-". Negative
 * amounts get a leading "-", e.g. `-Rp10.000,-`.
 *
 * Use the <MoneyText> / <CaPrice> components where a component fits; call this
 * directly for text-only spots (e.g. <option> labels, StatCard `sub` strings,
 * chart tooltips) where a component cannot be rendered.
 */
export function formatIDR(amount: number | null | undefined): string {
	const rounded = Math.round(Number(amount ?? 0));
	const safe = Number.isFinite(rounded) ? rounded : 0;
	const digits = new Intl.NumberFormat('id-ID', { maximumFractionDigits: 0 }).format(
		Math.abs(safe)
	);
	return `${safe < 0 ? '-' : ''}Rp${digits},-`;
}
